ldm/dream/args.py

```diff
@@ -0,0 +1,485 @@
+useful if both the args and the command contain the same attribute and
+you wish to apply logic as to which one to use. For example:
+
+  a = Args()
+  args    = a.parse_args()
+  opts    = a.parse_cmd(string)
+  do_grid = args.grid or opts.grid
+
+To add new attributes, edit the _create_command_arg_parser() and
+_create_cmd_parser() methods.
+
+We also export the function build_metadata
+"""
+
+import argparse
+import shlex
+import json
+
+SAMPLER_CHOICES = [
+    'ddim',
+    'k_dpm_2_a',
+    'k_dpm_2',
+    'k_euler_a',
+    'k_euler',
+    'k_heun',
+    'k_lms',
+    'plms',
+]
+
+# is there a way to pick this up during git commits?
+APP_ID      = 'lstein/stable-diffusion'
+APP_VERSION = 'v1.15'
+
+class Args(object):
+    def __init__(self):
+        '''Initialize new Args class. Takes no arguments'''
+        self._arg_parser   = self._create_arg_parser()
+        self._cmd_parser   = self._create_cmd_parser()
+        self._arg_switches = self.parse_args()   # so there's something there
+        self._cmd_switches = self.parse_cmd('')  # so there's something there
+
+    def parse_args(self):
+        '''Parse the shell switches and store.'''
+        self._arg_switches = self._arg_parser.parse_args()
+        return self._arg_switches
+
+    def parse_cmd(self,cmd_string):
+        '''Parse a dream>-style command string '''
+        command = cmd_string.replace("'", "\\'")
+        try:
+            elements = shlex.split(command)
+        except ValueError:
+            print(traceback.format_exc(), file=sys.stderr)
+            return
+        switches = ['']
+        switches_started = False
+
+        for el in elements:
+            if el[0] == '-' and not switches_started:
+                switches_started = True
+            if switches_started:
+                switches.append(el)
+            else:
+                switches[0] += el
+                switches[0] += ' '
+        switches[0] = switches[0][: len(switches[0]) - 1]
+        self._cmd_switches = self._cmd_parser.parse_args(switches)
+        return self._cmd_switches
+
+    def json(self,**kwargs):
+        return json.dumps(self.to_dict(**kwargs))
+
+    def to_dict(self,**kwargs):
+        a = vars(self)
+        a.update(kwargs)
+        return a
+
+    # Isn't there a more automated way of doing this?
+    # Ideally we get the switch strings out of the argparse objects,
+    # but I don't see a documented API for this.
+    def prompt_str(self,**kwargs):
+        """Normalized prompt."""
+        a = vars(self)
+        a.update(kwargs)
+        switches = list()
+        switches.append(f'"{a["prompt"]}')
+        switches.append(f'-s {a["steps"]}')
+        switches.append(f'-W {a["width"]}')
+        switches.append(f'-H {a["height"]}')
+        switches.append(f'-C {a["cfg_scale"]}')
+        switches.append(f'-A {a["sampler_name"]}')
+        switches.append(f'-S {a["seed"]}')
+        if a['seamless']:
```

Comment:
Python > 3.10 has [Structural Pattern Matching](https://peps.python.org/pep-0636) (the PEP is maybe over-long and over-detailed, but can be skimmed effectively) - think regex+'case'